### PR TITLE
Game material difference

### PR DESCRIPTION
**Pull request by ZTL-UwU**

### CODE REVIEW

**Comment by ZTL-UwU** [comment]
Will add the icons later

**Comment by veloce** [comment]
Hi thanks for this! A screenshot or video showing the feature would be nice.

**Comment by ZTL-UwU** [comment]
@veloce

File: lib/src/model/game/game.dart [hunk 1]
```diff
@@ -87,13 +90,98 @@ class ClockData with _$ClockData {
   }) = _ClockData;
 }
 
+@freezed
+class MaterialDiffSide with _$MaterialDiffSide {
+  const factory MaterialDiffSide({
+    required Map<Role, int> pieces,
```
Comment (veloce) [inline lib/src/model/game/game.dart:96]: You should use an `IMap` instead.

File: lib/src/model/game/game.dart [hunk 2]
```diff
@@ -87,13 +90,98 @@ class ClockData with _$ClockData {
...
+class MaterialDiff with _$MaterialDiff {
+  const MaterialDiff._();
+
+  const factory MaterialDiff({
+    required MaterialDiffSide black,
+    required MaterialDiffSide white,
+  }) = _MaterialDiff;
+
+  factory MaterialDiff.fromBoard(Board board) {
+    final Map<Role, int> pieceScores = {
```
Comment (veloce) [inline lib/src/model/game/game.dart:111]: should be defined as a `const IMap<Role, int>` outside

File: lib/src/ui/game/archived_game_screen.dart [hunk 3]
```diff
@@ -76,6 +81,41 @@ class ArchivedGameScreen extends ConsumerWidget {
   }
 }
 
+class _AppTitle extends ConsumerWidget {
```
Comment (veloce) [inline lib/src/ui/game/archived_game_screen.dart:84]: This shouldn't be there, it's not part of the material difference feature, so please remove it.

I'd like to review features/improvements separately, and I'm not sure to accept this change.

File: lib/src/ui/user/user_activity.dart
Comment (veloce) [inline lib/src/ui/user/user_activity.dart:1]: I don't understand the change in the whole file. Again this is not part of the material diff feature.

Maybe you fixed a bug or you just improved the code, I have no idea. So this should be part of another PR where you explain what is this change. Thanks.

File: lib/src/widgets/player.dart [hunk 4]
```diff
@@ -15,15 +19,35 @@ class BoardPlayer extends StatelessWidget {
...
   });
 
   final Player player;
   final Duration? clock;
   final bool? active;
+  final MaterialDiffSide? diff;
 
   @override
   Widget build(BuildContext context) {
+    final IList<Role> roleList = {
```
Comment (veloce) [inline lib/src/widgets/player.dart:33]: that should be a `const` defined outside. But you can actually use `Role.values` instead.

File: lib/src/widgets/player.dart [hunk 5]
```diff
@@ -15,15 +19,35 @@ class BoardPlayer extends StatelessWidget {
...
+      Role.king,
+      Role.queen,
+      Role.rook,
+      Role.bishop,
+      Role.knight,
+      Role.pawn,
+    }.toIList();
+
+    final Map<Role, IconData> iconByRole = {
```
Comment (veloce) [inline lib/src/widgets/player.dart:42]: should be a private const defined outside

File: lib/src/widgets/player.dart [hunk 6]
```diff
@@ -101,16 +152,37 @@ class PlayerRating extends StatelessWidget {
...
   final TextStyle? style;
 
   @override
   Widget build(BuildContext context) {
     final ratingStr =
         rating is double ? rating.toStringAsFixed(2) : rating.toString();
-    return Text(
-      '$ratingStr${provisional == true || deviation > kProvisionalDeviation ? '?' : ''}',
-      style: style,
+    return Row(
```
Comment (veloce) [inline lib/src/widgets/player.dart:163]: This has again nothing to do with material diff, please remove it.

**Review by veloce — COMMENTED** [review]
Thanks for the screenshots.

Please add some unit tests to test the logic that compute the material diff and the score.

**Comment by veloce** [comment]
Oops, I messed up with git and your pr was not cleanly merged with a commit. But it's merged.
